examples/cxx/event_based_communication/src/publisher.cpp

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) 2024 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache Software License 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+#include "iox2/publisher.hpp"
+#include "iox/duration.hpp"
+#include "iox2/file_descriptor.hpp"
+#include "iox2/listener.hpp"
+#include "iox2/node.hpp"
+#include "iox2/notifier.hpp"
+#include "iox2/sample_mut.hpp"
+#include "iox2/service_name.hpp"
+#include "iox2/service_type.hpp"
+#include "iox2/waitset.hpp"
+#include "pubsub_event.hpp"
+#include "transmission_data.hpp"
+
+#include <iostream>
+#include <utility>
+
+using namespace iox2;
+
+constexpr iox::units::Duration CYCLE_TIME = iox::units::Duration::fromSeconds(1);
+constexpr uint64_t HISTORY_SIZE = 20;
+
+///////////////////////////////////////////////
+/// START: EventBasedPublisher declaration
+///////////////////////////////////////////////
+
+// High-level publisher class that contains besides a publisher also a notifier and a listener.
+// The notifier is used to send events like `PubSubEvent::SentSample` or `PubSubEvent::SentHistory`
+// and the listener to wait for new subscribers.
+class EventBasedPublisher : public FileDescriptorBased {
```

Comment:
I know the Rust version is already in, but this name is really confusing to me. It sounds to me like a publisher implemented on top of some kind of event mechanism (e.g. like it is used in the terms "token-based authentication"), rather than one that augments publishing with notification.

If there is room to adapt this, I would suggest something more like `NotifyingPublisher` and `ListeningSubscriber` ?

Reply:
I will do rename them in the Rust and C++ example in this PR ... in 8 hours or later ;)